### ISSUE TITLE
Why are there so many execution methods for SpotifyRestAction?

### ISSUE THREAD

**Issue by Computr0n** [adamint/spotify-web-api-kotlin#221]
First off, great work on the library.

My question is regarding this design choice:

> ### What is the SpotifyRestAction class?
> Abstracting requests into a `SpotifyRestAction` class allows for a lot of flexibility in sending and receiving requests. 
> This class includes options for asynchronous and blocking execution in all endpoints. However, 
>  due to this, you **must** call one of the provided methods in order for the call 
>  to execute! The `SpotifyRestAction` provides many methods and fields for use, including blocking and asynchronous ones. For example,
>  - `hasRun()` tells you whether the rest action has been *started*
>  - `hasCompleted()` tells you whether this rest action has been fully executed and *completed*
> - `complete()` blocks the current thread and returns the result
> - `suspendComplete(context: CoroutineContext = Dispatchers.Default)` switches to given context, invokes the supplier, and synchronously retrieves the result.
> - `suspendQueue()` suspends the coroutine, invokes the supplier asynchronously, and resumes with the result
> - `queue()` executes and immediately returns
> - `queue(consumer: (T) -> Unit)` executes the provided callback as soon as the request 
> is asynchronously evaluated
> - `queueAfter(quantity: Int, timeUnit: TimeUnit, consumer: (T) -> Unit)` executes the 
> provided callback after the provided time. As long as supplier execution is less than the provided 
> time, this will likely be accurate within a few milliseconds.
> - `asFuture()` transforms the supplier into a `CompletableFuture` (only JVM)

My least favorite aspect of this library is how verbose this whole set of execution methods is and I don't see any benefit to it. 
No other library in the Kotlin ecosystem follows this pattern to my knowledge.

Couldn't the actions simply be suspend functions? (`suspendComplete()`)

All the other execution methods could be easily attained by using the corresponding coroutine methods/builders; we can already make a Future from a coroutine with an adapter, we can transform async actions into sync ones by `runBlocking { }`, we can queue them with actors or flows and we can detect when they complete by returning from either of these Async/Sync calls.
At the core of all these operations is something async, so why not just make them suspend funs and rely on the language to bridge out of async-land?

I think ideally all these actions would be simple `suspend fun`s, but If they need to be classes for some reason, the [classes could simply provide invoke overrides](https://try.kotlinlang.org/#/Kotlin%20Koans/Conventions/Invoke/Task.kt) with something like:
```kotlin
suspend operator fun invoke(parameters: P): Result<R>
``` 
to make them work like suspend funs in the typical usecase. 
The other execution methods could simply be provided as extension functions on a base SpotifyRestAction that are optionally in scope if they are really needed for some reason.

A good example of what I am suggesting here can be found in [Google's kotlin reference Android app IO-Schedule here with the UseCase class](https://github.com/google/iosched/blob/84415833c606210f8cddc1458ea23c02a8401d3c/shared/src/main/java/com/google/samples/apps/iosched/shared/domain/CoroutineUseCase.kt).

**Comment by adamint** [comment]
It's a good question, and one that has a bit of history behind it.

The short answer to your question "Couldn't the actions simply be suspend functions" is yes.
This library actually *precedes* the 1.0.0 release of Kotlin coroutines, and so most of these functions (queue, queueAfter, complete) were originally utilities to make this library easier to work with on JVM, modeled after the Discord wrapper (https://github.com/DV8FromTheWorld/JDA) that this library was originally developed to work with.

With that said, one of the goals of this library is to be flexible and provide some utility functions so that people don't necessarily need to "make a Future from a coroutine with an adapter," "transform async actions into sync ones by runBlocking { }," or "queue them with actors or flows" themselves, and can instead rely on this library to do that for them.

My thoughts are as follows:
- That this library provides several execution methods for different purposes is not a negative in my opinion, though if you feel otherwise, please do say so
- Providing invocation overrides is a good idea, and one that I had not thought of
- Following the extension method trend of major Kotlin libraries is a good idea, and could be done fairly easily

**Comment by Computr0n** [comment]
Good historical perspective.

> My thoughts are as follows:
> 
> * That this library provides several execution methods for different purposes is not a negative in my opinion, though if you feel otherwise, please do say so
> * Providing invocation overrides is a good idea, and one that I had not thought of
> * Following the extension method trend of major Kotlin libraries is a good idea, and could be done fairly easily

I think it is a negative since these invocation methods suggest to new users that thats what they should use. 
It also makes your library becomes responsible for maintaining and explaining them with documentation. 

Kotlin users of async operations should know how to work with suspend funs, and the coroutines libraries provides all of the necesary builders, adapters and so on in order for you to bridge out to the sync world without you having to rebuild the wheel.

**Comment by adamint** [comment]
I can see your argument for `queue`, `queueAfter`, `complete`, and `suspendQueue`. It could be confusing for users as to which to use. Deprecating these methods and introducing a suspend `execute` function would be okay for now. 

However, I don't think that removing any `SpotifyRestActionPaging` function would be useful.

**Comment by Computr0n** [comment]
You could follow the pattern that Google did when they moved Pager 2 to a pure Kotlin library with Pager 3.
They made the core library/dependency only have suspend funs and flows and provided another parallel/optional/additional dependency with RxJava/Future extension functions for those stuck on Java to be able to call them through RxJava or Guava:

https://developer.android.com/topic/libraries/architecture/paging/v3-overview

How they recommend consuming their library:
```groovy
dependencies { 
  implementation "androidx.paging:paging-runtime:$paging_version" 

  // optional - RxJava2 support
  implementation "androidx.paging:paging-rxjava2:$paging_version"

  // optional - Guava ListenableFuture support
  implementation "androidx.paging:paging-guava:$paging_version"
}
```

The benefit of following this approach is your library lives up to its name of Spotify-Web-API-**Kotlin** and the core library will not force you to depend on more than whats absolutely necessary. 

The app I am building with this is a pure Kotlin Android Instant App which has a file size limit of 4mb in order to work as an instant app. Depending on more things than I need increases complexity/build-times/file-sizes and so on. Every bit counts, less is more and all that...

Again, I want to say great work on the library though; none of this criticism is meant to take away from that.

**Comment by Computr0n** [comment]
Oh, and with regards to the SpotifyRestActionPaging, many of those can be done with flow operators directly on top of the base flow. things like reverse or flowing from some point in the middle. When I was hooking your library up to Pager 3 on android, I basically had to make an adapter that did that anyway, none of the existing ones worked lazily as per Pager 3's requirements...

**Comment by adamint** [comment]
A `spotify-api-kotlin-java-ext` artifact could provide java extensions for what you're asking. This would be a good project to do. However, if you're looking to get a smaller core artifact, that's likely not going to be able to happen. Ktor client, kotlinx.serialization, and kotlinx.coroutines are the only dependencies for JVM, with a few other necessary dependencies for JS/Android integration.

**Comment by adamint** [comment]
@Computr0n is this something you would want to take the lead on? I will be able to, however my bandwidth is reduced for a few weeks due to work

**Comment by adamint** [comment]
@Computr0n this is resolved with #228
